src/ring_theory/congruence.lean

```diff
@@ -271,4 +276,167 @@ def mk' [non_assoc_semiring R] (c : ring_con R) : R →+* c.quotient :=
 
 end quotient
 
+/-! ### Lattice structure
+
+The API in this section is copied from `group_theory/congruence.lean`
+-/
+
+section lattice
+variables [has_add R] [has_mul R]
+
+/-- For congruence relations `c, d` on a type `M` with multiplication and addition, `c ≤ d` iff
+`∀ x y ∈ M`, `x` is related to `y` by `d` if `x` is related to `y` by `c`. -/
+instance : has_le (ring_con R) := ⟨λ c d, ∀ ⦃x y⦄, c x y → d x y⟩
+
+/-- Definition of `≤` for congruence relations. -/
+theorem le_def {c d : ring_con R} : c ≤ d ↔ ∀ {x y}, c x y → d x y := iff.rfl
+
+/-- The infimum of a set of congruence relations on a given type with multiplication and
+addition. -/
+instance : has_Inf (ring_con R) :=
+⟨λ S,
+  { r := λ x y, ∀ c : ring_con R, c ∈ S → c x y,
+    iseqv := ⟨λ x c hc, c.refl x, λ _ _ h c hc, c.symm $ h c hc,
+              λ _ _ _ h1 h2 c hc, c.trans (h1 c hc) $ h2 c hc⟩,
+    add' := λ _ _ _ _ h1 h2 c hc, c.add (h1 c hc) $ h2 c hc,
+    mul' := λ _ _ _ _ h1 h2 c hc, c.mul (h1 c hc) $ h2 c hc }⟩
+
+/-- The infimum of a set of congruence relations is the same as the infimum of the set's image
+    under the map to the underlying equivalence relation. -/
+lemma Inf_to_setoid (S : set (ring_con R)) : (Inf S).to_setoid = Inf (to_setoid '' S) :=
+setoid.ext' $ λ x y, ⟨λ h r ⟨c, hS, hr⟩, by rw ←hr; exact h c hS,
+  λ h c hS, h c.to_setoid ⟨c, hS, rfl⟩⟩
+
+/-- The infimum of a set of congruence relations is the same as the infimum of the set's image
+    under the map to the underlying binary relation. -/
+lemma Inf_def (S : set (ring_con R)) :
+  ⇑(Inf S) = Inf (@set.image (ring_con R) (R → R → Prop) coe_fn S) :=
+by { ext, simp only [Inf_image, infi_apply, infi_Prop_eq], refl }
+
+instance : partial_order (ring_con R) :=
+{ le := (≤),
+  lt := λ c d, c ≤ d ∧ ¬d ≤ c,
+  le_refl := λ c _ _, id,
+  le_trans := λ c1 c2 c3 h1 h2 x y h, h2 $ h1 h,
+  lt_iff_le_not_le := λ _ _, iff.rfl,
+  le_antisymm := λ c d hc hd, ext $ λ x y, ⟨λ h, hc h, λ h, hd h⟩ }
+
+/-- The complete lattice of congruence relations on a given type with multiplication and
+addition. -/
+instance : complete_lattice (ring_con R) :=
+{ inf := λ c d,
+  { to_setoid := (c.to_setoid ⊓ d.to_setoid),
+    mul' := λ _ _ _ _ h1 h2, ⟨c.mul h1.1 h2.1, d.mul h1.2 h2.2⟩,
+    add' := λ _ _ _ _ h1 h2, ⟨c.add h1.1 h2.1, d.add h1.2 h2.2⟩ },
+  inf_le_left := λ _ _ _ _ h, h.1,
+  inf_le_right := λ _ _ _ _ h, h.2,
+  le_inf := λ _ _ _ hb hc _ _ h, ⟨hb h, hc h⟩,
+  top := { mul' := λ _ _ _ _ _ _, trivial,
+           add' := λ _ _ _ _ _ _, trivial,
+           ..(⊤ : setoid R) },
+  le_top := λ _ _ _ h, trivial,
+  bot := { mul' := λ _ _ _ _, congr_arg2 _,
+           add' := λ _ _ _ _, congr_arg2 _,
+           ..(⊥ : setoid R) },
+  bot_le := λ c x y h, h ▸ c.refl x,
+  .. complete_lattice_of_Inf (ring_con R) $ assume s,
+    ⟨λ r hr x y h, (h : ∀ r ∈ s, (r : ring_con R) x y) r hr, λ r hr x y h r' hr', hr hr' h⟩ }
```

Comment:
Can you first get `distrib_lattice (ring_con R)` by pulling back from ,`R → R → Prop`? See `simple_graph.distrib_lattice` for what I mean precisely.